pkg/maps/metricsmap/metricsmap.go

```diff
@@ -86,6 +133,64 @@ var (
 		})
 )
 
+// updatePrometheusMetrics checks the metricsmap key value pair
+// and determines which prometheus metrics along with respective labels
+// need to be updated.
+func updatePrometheusMetrics(key *Key, val *Value) {
+	var counter prometheus.Counter
+	var err error
+	if key.IsDrop() {
+		counter, err = metrics.DropCount.GetMetricWithLabelValues(key.DropForwardReason(), key.Direction())
+	} else {
+		counter, err = metrics.ForwardCount.GetMetricWithLabelValues(key.Direction())
+	}
+	if err != nil {
+		log.WithError(err).Warn("failed to update prometheus metrics")
+		return
+	}
+	oldValue := metrics.GetCounterValue(counter)
+	newValue := val.CountFloat()
+	// Check if metrics have changed since the last poll.
+	// If yes, we need to add only the delta.
+	if newValue > oldValue {
+		if key.IsDrop() {
+			metrics.DropCount.WithLabelValues(key.DropForwardReason(), key.Direction()).Add((newValue - oldValue))
+		} else {
+			metrics.ForwardCount.WithLabelValues(key.Direction()).Add((newValue - oldValue))
+		}
+	}
+}
+
+// SyncMetricsMap is called periodically to sync of the metrics map by
+// aggregating it into drops (by drop reason), drops (by direction) and
+// forwards (by direction) with the prometheus server.
+func SyncMetricsMap() error {
+	file := bpf.MapPath(MapName)
+	metricsmap, err := bpf.OpenMap(file)
+
+	if err != nil {
+		return fmt.Errorf("unable to open metrics map: %s", err)
+	}
+	defer metricsmap.Close()
+
+	var key, nextKey Key
+	for {
+		err := metricsmap.GetNextKey(&key, &nextKey)
```

Comment:
Is there a particular reason not to reuse `Map.Dump()` / `Map.DumpWithCallback()` and the `Metrics` map instance in this package for iterating the map / updating prometheus metrics?

Reply:
Thanks for the review Joe!. `Map.Dump() / Map.DumpWithCallback()` we mostly used in the `cmd` files. I guess I was seeing the various ways in this code in `pkg/maps/*maps` were iterating bpf maps and did no come across `Map.Dump`. This is definitely an option, but in the interest of time I will refrain from making this change as part of this PR, because this is more of a style change as opposed to correctness of code. I can do it as part of a followup PR. Thanks!

Reply:
If the map is being dumped without these `Dump` and `DumpWithCallback` functions, you should `RLock` the map which is done in `pkg/bpf/map/DumpWithCallback` to ensure that it is not modified from userspace while accessing it. I don't think this is an issue now because it's only updated in the datapath, but it could be changed in the future.